add playground for example/tutorial step-92

webwurli opened the pull request:
As announced in #17836, here's some material to discuss for a tutorial step-92 concept.
Appreciate your feedback. ;-)

webwurli committed:
add playground for example/tutorial step-92

webwurli committed:
tutorial step-92: made indent and changed revised typos

webwurli committed:
version 0.2 of step-92 example/tutorial

webwurli committed:
added starting point for doxygen documentation - work in progress

webwurli committed:
indentation fixed

webwurli committed:
documentation: adopted kind of tutorial

webwurli committed:
doxygen: reveiwed double sections

webwurli commented:
I just removed the python scripts and provide a gmsh example-1.geo file instead.
It's the two-conductor-example from the workshop in August '24 (extended by an optionally conductive or dielectric solid in between).

Any wishes to keep the python-based model-generation instead of the example-1.geo ?? (I also improved the python files.)

Generate the mesh with gmsh:  "gmsh example-1.geo -3"

The solver is based on MUMPS. At start, it will load the parameters.prm file. On exit, it exports data which can be visualized using Paraview. (See comment below)
This rudimentary solver may be replaced by a more elegant one. I recommend to take the parameter reading and the CellsPostprocessor from this one.

The parameters.prm file contains assignments of material properties and boundary values as well as the mesh file name and the cell data output file name (both without extensions).

The paraview state file can be used for visualization but most likely, the solution_cells.vtu has to be selected manually as the state-file has absolute path names.

The documentation is at its very beginning: ...work in progress.

webwurli committed:
fixed Assert(); fixed debug make

webwurli committed:
created tests/examples/step-92.diff

webwurli committed:
created tests/examples/step-92.diff

webwurli committed:
tutorial documentation - 1st. draft

webwurli committed:
added example-2: a very very simple and theoretical two-winding transformer

webwurli committed:
removed obsolete file example-1_field_plot_1.png --> it has been replaced by a jpg-file

webwurli committed:
fixed example-2 model key and documentation

webwurli committed:
flipped orientation of winding_1 in example-2

jppelteret commented:
Hi @webwurli, 

Sorry that its taken me (or anyone else) quite some time to contribute anything to your PR. I took a skim through it and I think that it is really promising. I don't think that we have a tutorial with the electrostatics formulation that you've implemented, so between that and the complex geometries I think that this is going to be a great contribution. I really like the two examples that you've shared. The results look awesome! And I think that you did the right thing by removing the python generation script -- that's probably too much for a tutorial. If you wanted to, this could be shared as a part of a further code gallery problem if you wanted to go that route.

Before adding more detailed comments on the PR itself, can I ask if you've considered choosing either the `p::d::Triangulation` or the `p::fully_dist::Triangulation`, rather than having a switch between them? Or do you think that there's a particular point to showing the users how to set up one or the other? Do you think that they both have merit when solving on complex geometries, or do you tend to gravitate towards one or the other in practice?

Best,
J-P

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+ * the top level directory of deal.II.
+ *
+ * ---------------------------------------------------------------------
+ */
```
We'll have to restrict this to the standard headers for tutorials only, I think. But this can be done at the end of the review.

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+  double sum(const double &a, const double &b)
+  {
+    return a + b;
+  };
```
I believe these `sum` functions are unused, and can be removed.

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+  {
+  public:
+    ProblemParameters();
+    ~ProblemParameters();
```
```suggestion
    ~ProblemParameters() = default;
```
and you can then delete the implementation that follows later.

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+// associated with (non-negative) integral indices $[0,N)$, then both the set
+// of elements the current processor owns as well as the (possibly larger) set
+// of indices it needs to know about are subsets of the set $[0,N)$. IndexSet
+// is a class that stores subsets of this set in an efficient format:
```
We'll have to decide which headers are worth expanding on in the form of added documentation, and which are already covered in other tutorials. (We can also point users to the other tutorials when necessary.)

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+
+  public:
+    unsigned long        N_physicals = 0;
+    material_definition *p_physicals = NULL;
```
```suggestion
    material_definition *p_physicals = nullptr;
```

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+    material_definition *p_physicals = NULL;
+
+    unsigned long       N_boundary_values = 0;
+    boundary_potential *p_boundary_values = NULL;
```
```suggestion
    boundary_potential *p_boundary_values = nullptr;
```

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+          }
+      }
+
+    return p_physicals[default_physical_ix];
```
I don't think that you'd expect/want to hit this line, so maybe add an assertion beforehand that triggers a failure if you reach it? Similar for the final return in `ProblemParameters::get_boundary_value()` below.

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+    using BoundaryPotential_tuple =
+      std::tuple<std::string, physical_ID, double>;
+
+    class material_definition
```
You probably want to stick with a consistent scheme for naming classes, types, etc. I think that `class MaterialDefinition` and `using FieldVector_t = ...` would be typical for tutorials (but we should check to be sure). To keep things simple (+ consistent with the usual style of tutorials), you could also remove some of the type definitions. The `physical_ID` is OK in my opinion, I think that the interpretation of the `Tensor` -> Field vector type would be understood by most, and so is not necessary.

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+    std::string get_mesh_filename();
+    std::string get_cells_solution_filename();
+
+    // double mu_vacuum;      // vacuum permeability: 4 pi 10e-7 V s / (A m)
```
Looks like this can be removed?

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+      std::string        name;
+      double epsilon; // absolute electric permittivity epsilon (not relative !!
+                      // --> eps_r * eps_0)
+      double kappa;   // absolute electric conductivity
```
FYI, the tactical comments would ultimately be moved to some general class documentation for this `material_definition` class. Similar for the `boundary_potential`.

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+    std::string get_cells_solution_filename();
+
+    // double mu_vacuum;      // vacuum permeability: 4 pi 10e-7 V s / (A m)
+    double epsilon_vacuum; // vacuum permittivity: 8.8541878188(14)×10^−12 A s /
```
You can make this `constexpr` and define inline:
```
static constexpr double epsilon_vacuum = 8.8541878176204199e-12; // [ A s / (V m) ]
```

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+                  boundary_values_list,
+                  "list of potential definitions: \"name : ID : epot [V]\"");
+  }
+
```
I think that we typically have two new lines between function definitions.

jppelteret commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,1091 @@
+
+        ipot++;
+      }
+  }
```
I'm sure that you planned to do it, but it would be worth sketching out what these operations in the call to `finalize()` are doing.

jppelteret reviewed (commented):
Initial batch of (general) comments covering the preliminaries and `ProblemParameters`.

webwurli committed:
update example 2 geometry and plots

webwurli committed:
cleaned up source-code, geo-files, prm-files, removed pvsm-files and improved documentation

webwurli committed:
cleaned up source-code, geo-files, prm-files, removed pvsm-files and improved documentation

webwurli commented:
> Initial batch of (general) comments covering the preliminaries and `ProblemParameters`.

Hi JP.

Thank You for all your hints.

I'm actually trying to push my changes, but when I want to push my feature branch
`git push origin add_example_step_92`
it's rejected because the tip of my current branch is behind its remote counterpart.

(p.s.: I already rebased master.)

webwurli commented:
I'd need some advise how to fix this PR (or alternately close it and create a new one?)
Thank You.

p.s.:
Seems I messed it up with a rebase - I apologize.
This tutorial draft shall only contain
  6 files in folder "./examples/step-92/"
  12 files in folder "./examples/step-92/doc/" and

marcfehling commented:
Hi @webwurli,

it looks like something went wrong rebasing indeed. The master branch of your deal.II repository on github is up-to-date, but the feature branch for this PR forked off about 1,000 commits ago.

On your local repository on your laptop, I suggest you switch to your `master` branch, and pull the changes from your remote github repository. Then switch to your feature branch (`add_example_step_92`) again, try `git rebase master`, and then force push to your remote feature branch.

I tried it out on my laptop by pulling your feature branch directly. As an alternative and only if you'd like, I offer to push to your branch directly and fix it right away.

webwurli committed:
improved consistency of naming and numbering

webwurli committed:
deleted unused jpg pictures

webwurli committed:
fixed indentation

marcfehling commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,981 @@
+  // The solution itself is the electrostatic potential.
+  // In order to export it, the <code>update_values</code> flag is needed.
+  // For plotting electric streamlines or currents, the gradients of the
+  // solution will be neded and hence, the <code>update_gradients</code> has to
```
```suggestion
  // solution will be needed and hence, the <code>update_gradients</code> has to
```

marcfehling commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,981 @@
+  //     ".msh") containing the triangulation
+  //   </li>
+  //   <li>
+  //     "cells solution filename" is sepcifying filename (without suffix) for
```
```suggestion
  //     "cells solution filename" is specifying filename (without suffix) for
```

marcfehling commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,409 @@
+    <td>
+      <img src="../../../examples/step-92/doc/example-1_model_key.jpg" width="50%" alt="example-1, 3D model"/>
+      <p> example-1, 3D model: ohmic conductors (green), terminals (light blue), barrier (blue)</p>
+
```
```suggestion
```
Remove this empty line.

marcfehling commented on examples/step-92/doc/results.dox:
```diff
@@ -0,0 +1,497 @@
+
+<h3> example-2 </h3>
+
+<h4> geometry and parameters </h4>
```
You have used this header before in line 5 of this file. It seems like doxygen doesn't like that: it is creating section labels for the html version based on the title, and assumes they are unique. Can you try a different name?

marcfehling reviewed (commented):
This should take care of most of the warnings of the `indent` CI.

Can you please add a file `tests/examples/step-92.with_umfpack=true.output` that contains the console output of your program in textform? This way, your tutorial will be picked up by our testsuite.

webwurli committed:
corrections as proposed

webwurli commented:
Thank You @marfehling. That was very helpful.

I also added that "tests/examples/step-92.with_umfpack=true.output", 
but I have doubts if a test-script will successfully run:
A mesh has to be generated before calling the solver.
(It's described in the tutorials documentation doc/results.dox.)

webwurli committed:
fixed indentation in documentation

marcfehling commented:
> I have doubts if a test-script will successfully run: A mesh has to be generated before calling the solver.

That is exactly what the `.diff` files are for. See also [tests/examples/README.md](https://github.com/dealii/dealii/blob/master/tests/examples/README.md)

We can add a few lines to your program that will call the gmsh API to generate the msh file. I have not much experience with gmsh. How would such a code snippet look like? Something like:
```
#include <gmsh.h>

void generate_msh_from_geo(const int dim)
{
  gmsh::initialize(argc, argv)
  gmsh::open("example-1.geo");
  gmsh::model::mesh::generate(dim);
  gmsh::write("example-1.msh");
  gmsh::finalize();
}
```

marcfehling commented:
@webwurli -- How large is the `msh` file (in bytes)? If it's small, we can also just add it to the `step-92` folder.

@tamiko -- Do you see a different way how we can call `gmsh` prior to running the tutorial test?

marcfehling commented:
@webwurli -- I installed gmsh and created `example 1.msh`. However, `step-92` crashes with
`mesh_filename.size() > 0` in Debug mode and `Assertion '__n < this->size()' failed.` in Release mode. Can you double check?

webwurli commented:
> @webwurli -- How large are the `msh` files (in bytes)? If they're small, we can also just add them to the `step-92` folder.
> 
> @tamiko -- Do you see a different way how we can call `gmsh` prior to running the tutorial test?

```
  1100002 Apr  3 06:47 example-1.msh
 20137577 Apr  3 06:47 example-2.msh
```

webwurli committed:
fixed debug mode Assert crashes and implemented gmsh generate in make_grid function

webwurli commented:
> @webwurli -- I installed gmsh and created `example 1.msh`. However, `step-92` crashes with `mesh_filename.size() > 0` in Debug mode and `Assertion '__n < this->size()' failed.` in Release mode. Can you double check?


Thanks for the hint.....yes, the Assert after reading the parameter-file mas placed in wrong location.
That cause the crash in debug-mode.

I changed it. But the crash in release-mode I cannot reproduce (even before that change). It doesn't crash on my system. (-->??<--)

I also implemented your suggestion to call gmsh in step-92. 
So the recently committed version of step-92 (at least on my system) ran in 4 variants:

debug - with or without existing .msh
release - with or without existing .msh


```
make debug;
make
rm *.msh
./step-92
./step-92
make release;
make
./step-92
./step-92
```

marcfehling committed:
Changed dependencies to Gmsh and PETSc.

marcfehling committed:
Use std::filesystem and MPI_Barrier for the gmsh part.

marcfehling committed:
Fixed test/examples version.

marcfehling commented:
Thanks for fixing the assertion. I pulled your branch again and now it works in both Debug and Release mode. I guess your changes took care of it. Also thank you for adding the gmsh part.

I pushed some changes to your branch to make the step-92 test work.

webwurli commented:
I've been searching for the MPI Barrier in the deal.II Utilities::MPI 
....you fixed my dirty trick ;-)  ....thanks!   ( So MPI_Barrier is directly usable :+1:  )

webwurli committed:
for testsuite: suppress gmsh output and step-92's printing number of degrees of freedom

webwurli commented:
oh no! .....again!
...wait a moment....I'll fix

webwurli commented:
changed the step-92.diff such that only for testsuite verbosity of step-92  is reduced:

1. gmsh verbosity is set to 1
2. number of degrees of freedom shall not be printed during setup()

webwurli committed:
fixed tiping error in testsuite .diff file

webwurli commented:
@marcfehling -- now I see you already did that .diff file.  I apologize I've damaged your commit
`Fixed test/examples version. [b4dcfc2]`

webwurli committed:
aligned fix of tests/examples version co commit b4dcfc2

webwurli commented:
Still have to

- link step-92 in the "doc/doxygen/tutorial/tutorial.h.in" and
- maybe add some more interpretation of examples results in doc/results.dox

webwurli committed:
added entry in tutorial overview page

webwurli committed:
small fixes in tutorial step-92 documentation

webwurli commented:
What do I have to do for starting those 4 pending checks "Jenkins: ampere, MPI, OSX and serial" ?

masterleinad commented:
/rebuild

webwurli committed:
Merge branch 'master' into add_example_step_92

tjhei reviewed (commented):
I think this is a very neat example. My original idea when talking to you about this was to figure out a way to handle very large external meshes efficiently. I think you are currently still loading the whole mesh redundantly on every MPI rank.
A nice addition (maybe after merging this program?) would be to optionally partition externally and only loading the "parts" directly.

tjhei commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+// importing meshes when using parallel::distributed
+// vs. parallel::fullydistributed triangulation.
+//
+// If the option USE_FULLY_DISTRIBUTED_TRIA is defined,
+// the solver will be compiled for using
+// parallel::fullydistributed triangulation
+// otherwise, parallel::distributed triangulation will be used.
```
What is the motivation of this switch?

webwurli committed:
Update examples/step-92/step-92.cc

correction of typo - thanks

Co-authored-by: Timo Heister <timo.heister@gmail.com>

bangerth commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,422 @@
+
+The efficiency of the shown system is $\eta = \frac{P_2}{P_1} < 1.0$.
+When designing this unrealistically simplified transmission line
+for a combination of maximum efficiency and reasonable invest,
```
```suggestion
for a combination of maximum efficiency and reasonable investment,
```

bangerth commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,422 @@
+In large (national or international) power transmission systems,
+this optimization usually leads to systems with high and ultra-high voltages
+and many more aspects have to be considered, i.e. reactive power,
+failure modes, or dc vs. ac transmission.
```
```suggestion
failure modes, or DC (direct current) vs. AC (alternating current) transmission.
```

bangerth commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,422 @@
+system voltages leads to the fact that all connected equipment
+has to be designed for safe and economic operation at those voltages.
+
+In design of high voltage equipment, it is common
```
```suggestion
In the design of high voltage equipment, it is common
```

bangerth commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,422 @@
+      <p>
+          Two ohmic conductors, each connected with two terminals.
+          Current flow inside the conductors can be enforced by defining
+          different electric potential at each terminal.
```
```suggestion
          different electric potentials at each terminal.
```

bangerth commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,422 @@
+<ul>
+  <li>
+  Dielectric material properties show linear behaviour.
+  (In this tutorial, dielectric permittivity is assumed to be lossless $\permittivity\in\mathbb{R}$.)
```
I had to think about what it is you were trying to say withe formula at the end of the line until I realized that in essence you mean that the permittivity is *not* complex-valued. Perhaps it's easiest to say it that way:
```suggestion
  (In this tutorial, dielectric permittivity is assumed to be lossless, which is typically modeled by using a real-valued -- rather than complex-valued -- permittivity $\permittivity$.)
```

bangerth commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,422 @@
+        Inside conductors, current density is much higher than dielectric displacement current:
+        $ \left\| \cvJelfield \right\| \gg \omega\permittivity || \cvEfield || $
+
+        Example for copper conductivity
```
```suggestion
        For example, for copper conductivity
```

bangerth commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,422 @@
+    </li>
+    <li>
+        For higher frequencies, Laplace equation will not serve to solve the problem anyhow as
+        induction and wave-propagation are completely neglected when not solving the full Maxwell's equations.
```
I think it might be worthwhile adding a short discussion about what low/high frequencies are here. Perhaps something along the following lines (use your own words, of course, if you think that my explanation is not good enough):

At the core of the question whether one can use the Laplace equation or needs to consider a time-dependent model like Maxwell's equations is how far the electric and magnetic fields propagate in one alternating current (AC) cycle. For example, at $f=60 Hz$, using the speed of light of approximately $c=300,000 km/s$, fields can propagate 5000 km per cycle -- a distance far larger than the devices we want to consider; as a consequence, it is reasonable to assume that at every time point in the DC cycle, the fields in the vicinity of our device are as they would be if we considered a time-independent problem.  In other words, we can use the Laplace equation as long as $L \ll \frac{c}{f}$, which is the case for the examples we consider in this tutorial program. Of course, there may be other physical effects that might requires us to consider more complex models than the Laplace equation.

bangerth commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,422 @@
+  \\
+  \Epot &= g \qquad\qquad & \text{on}\ \partial\Omega
+@f}
+with a real-valued parameter $a$ for consideration of material properties (either conductor or insulator).
```
```suggestion
with a real-valued and spatially variable parameter $a$ for consideration of material properties (either conductor or insulator).
```

bangerth commented on examples/step-92/doc/intro.dox:
```diff
@@ -0,0 +1,422 @@
+        Terminals will be modeled as cavities and on its surfaces $\Gamma_{k}$
+        Dirichlet boundary conditions will be used for assigning electric potential values ${\Epot}_{k}$.
+    </li>
+</ol>
```
I think it would be useful to extend the introduction by a couple more sections:

* The equations you are solving are already solved in step-5 and step-6. Can you discuss what is different here? (Specifically: complicated geometry, parallel solution, ...)
* One of the key issues you had to deal with was the generation of meshes and how to read them in. Could you discuss what that involved?

In essence, I think of the introduction as "setting the scene" for reading the program. This is where you lay out what it is you will have to address in code in the section that follows the introduction (namely, the commented program) so that a reader already knows what will be discussed there.

bangerth reviewed (commented):
@webwurli First, my apologies for letting this sit for so long without further feedback! Let us see that we can get this program over the finish line before the next release this summer!

In the following are comments on the introduction. I will work my way through the remainder of it all over the course of the next several days. Thank you for working with us on getting this program into deal.II!

bangerth commented on examples/step-92/doc/results.dox:
```diff
@@ -0,0 +1,494 @@
+</table>
+
+The geometry does not contain any material properties nor boundary value definitions.
+For understanding how this will be done in this tutorial, a look at a list of the models physical groups in the gmsh GUI is useful.
```
What is "this" in the second sentence? I think what you want to say is "how using these material and boundary properties in the simulation".

```suggestion
The geometry does not contain any material properties nor boundary value definitions.
For understanding how this will be done in this tutorial, a look at a list of the model's physical groups in the gmsh GUI is useful.
```

bangerth commented on examples/step-92/doc/results.dox:
```diff
@@ -0,0 +1,494 @@
+Next section shows how to generate the mesh without opening the gmsh GUI,
+how to start the simulation and a graphical visualization of the output data.
+
+<h3> example-1: results </h3>
```
If the part above moves to the introduction, then this would be where the results section would start.

bangerth commented on examples/step-92/doc/results.dox:
```diff
@@ -0,0 +1,494 @@
+
+<h3> example-1: results </h3>
+
+Generate a coarse mesh in binary file format, run the program and see the output
```
```suggestion
In order to run the program on the first of the examples, let us first call `gmsh` to
generate a coarse mesh in binary file format, and then run the program and see the output:
```

bangerth commented on examples/step-92/doc/results.dox:
```diff
@@ -0,0 +1,494 @@
+</table>
+
+
+<h3> example-2: geometry and parameters </h3>
```
same here: I think this should go into the introduction.

bangerth commented on examples/step-92/doc/results.dox:
```diff
@@ -0,0 +1,494 @@
+@endcode
+@note
+Usually refinement shall be done within the solver and not
+for the initial mesh. Only, the example-2 is still to small for seeing
```
Do you mean to say that *typically*, we deal with mesh refinement within the finite element program rather than the mesh generator?

```suggestion
Usually refinement shall be done within the solver and not
for the initial mesh. Only, the example-2 is still too small for seeing
```

bangerth commented on examples/step-92/doc/results.dox:
```diff
@@ -0,0 +1,494 @@
+@endcode
+
+
+Following table shows a comparison of wall-times for the two
```
```suggestion
Following table shows a comparison of wall times for the two
```

bangerth reviewed (commented):
Following are comments on the results section. 

I often find those programs the most interesting where the results section does not just say "This is how you run the program and here is a picture of the output", but where the results section also has a discussion of what one actually learns from the outcome. It took me many years of writing tutorials before I realized this point and then wrote a long section in step-4 about this, https://dealii.org/developer/doxygen/deal.II/step_4.html#step_4-PostprocessingWhattodowiththesolution . Is there anything that you, as a professional in the field, took away from the simulation results that would be worth sharing?

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+        domain_index++;
+      }
+
+    // if no default domain properties have been defined in the parameter file,
```
```suggestion
    // If no default domain properties have been defined in the parameter file,
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+            gmsh::open(geofilename);
+            gmsh::model::mesh::generate(dim);
+            gmsh::write(mshfilename);
+            gmsh::finalize();
```
I think you can only call these if deal.II has been configured with gmsh. So you have to surround it with something like `#if DEAL_II_WITH_GMSH` and have some kind of error for the `#else` branch.

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+    const unsigned int mpi_size =
+      Utilities::MPI::n_mpi_processes(mpi_communicator);
+
+    auto construction_data = TriangulationDescription::Utilities::
```
I suspect you can make that `const`.

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+  // In case the mesh file does not exist, we call gmsh to generate it from the
+  // corresponding geometry file. We do this only on the very first MPI process,
+  // thus we need to ensure that all other processes wait until the mesh file is
+  // generated.
```
```suggestion
  // corresponding geometry file. We do this only on the very first MPI process,
  // thus we need to ensure that all other processes wait until the mesh file is
  // generated via an MPI barrier that ensures that all processes can only proceed
  // past it when all processes have reached the point of the barrier.
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+                    fe_face_values.reinit(cell, face);
+
+                    for (unsigned int q_point = 0; q_point < n_face_q_points;
+                         q_point++)
```
```suggestion
                    for (const unsigned int q_point : fe_face_values.quadrature_point_indices())
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+
+  // @sect4{Data Output}
+  // After solving the problem, some output data shall be written into a
+  // vtk-file. The output-filename has been specified in the parameter file and
```
```suggestion
  // vtk file. The output filename has been specified in the parameter file and
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+  // vtk-file. The output-filename has been specified in the parameter file and
+  // shall contain not only the solution of the Laplace PDE, but also electric
+  // field strengths, current densities and domain properties for each
+  // evaluation point. For this purpose a Cells Postprocessor is used.
```
```suggestion
  // evaluation point. For this purpose a data postprocessor is used that hooks into the DataOut
  // class and for each cell generates the data we want to output.
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+
+  // Define the interpretation of the variables in the output
+  // - The electrostatic potential is a scalar value
+  // - Electric field strength and current density are vectors
```
```suggestion
  // - Electric field strength and current density are vectors.
  //    Because this program always runs in 3d, each vector has
  //     three components.
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+    std::vector<Vector<double>>                &computed_quantities) const
+  {
+    unsigned int n_evaluation_points;
+    n_evaluation_points = inputs.solution_values.size();
```
```suggestion
    const unsigned int n_evaluation_points = inputs.solution_values.size();
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+        for (unsigned int d = 0; d < dim; d++)
+          {
+            solution_E_field[d] = -solution_gradient[d];
+          }
```
Would this work?
```suggestion
          solution_E_field = -solution_gradient;
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+  {
+    TimerOutput::Scope t(computing_timer, "05. post-process");
+
+    Assert(problem_parameters.cells_output_data_filename.size() > 0,
```
```suggestion
    Assert(problem_parameters.cells_output_data_filename != "",
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+    Assert(problem_parameters.cells_output_data_filename.size() > 0,
+           ExcInternalError());
+
+    std::string cells_output_data_filename =
```
```suggestion
    const std::string cells_output_data_filename =
```

bangerth commented on examples/step-92/step-92.cc:
```diff
@@ -0,0 +1,995 @@
+
+  // @sect4{Main algorithm}
+
+  // This is the central starting-point to run all simulation steps
```
```suggestion
  // This is the central starting-point to run all simulation steps:
```

bangerth reviewed (commented):
Here is the rest. It's a lot of small comments, but there is nothing big among them -- I think the program is actually in pretty good shape already!

Having read through it all, I think the following is really all that I'm missing:
* The program is more or less a copy of step-40, with the exception that it reads a complex mesh and that it uses a direct solver. I think it would be nice to discuss these two points in the introduction.
* That is, could you add a section that discusses the workflow that results in the .geo file that is then meshed?
* And could you discuss the choice of the solver? People will generally assume that multigrid-type algorithms are unbeatable. I assume that you tested this and came to the conclusion that the direct solver is faster. That would make for a good discussion -- share these sorts of insights with your readers!
* There is still the issue of the distributed vs fully distributed mesh. Let us postpone this for the moment -- we'll come back to it.

Thank you again for working on this. It will be fun to see this merged!

bangerth commented:
@webwurli Any news?

bangerth commented:
@webwurli Just checking in -- any news?

webwurli commented:
@bangerth 
I apologize for the delay. I'll not be able finalize it in a short timeframe. (I appreciate all your editorial hints. I'll consider them. I admit I've been distracted for a while.)

bangerth commented:
All good. I'll take it out of the list for the next release for now, but in the end we'll be happy to take your program any time you have the time to complete it!
